mava/systems/q_learning/sebulba/rec_iql.py

```diff
@@ -0,0 +1,803 @@
+# Copyright 2022 InstaDeep Ltd. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copy
+import queue
+import threading
+import warnings
+from collections import defaultdict
+from queue import Queue
+from typing import Any, Dict, List, Sequence, Tuple
+
+import chex
+import hydra
+import jax
+import jax.lax as lax
+import jax.numpy as jnp
+import numpy as np
+import optax
+from colorama import Fore, Style
+from flax.core.scope import FrozenVariableDict
+from flax.linen import FrozenDict
+from jax import Array, tree
+from jax.experimental import mesh_utils
+from jax.experimental.shard_map import shard_map
+from jax.sharding import Mesh, NamedSharding, PartitionSpec, Sharding
+from omegaconf import DictConfig, OmegaConf
+from rich.pretty import pprint
+
+from mava.evaluator import get_sebulba_eval_fn as get_eval_fn
+from mava.evaluator import make_rec_eval_act_fn
+from mava.networks import RecQNetwork, ScannedRNN
+from mava.systems.q_learning.types import Metrics, QNetParams, Transition
+from mava.systems.q_learning.types import SebulbaLearnerState as LearnerState
+from mava.types import Observation, SebulbaLearnerFn
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.config import base_sebulba_checks as check_sebulba_config
+from mava.utils.config import check_total_timesteps
+from mava.utils.jax_utils import switch_leading_axes
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.sebulba.pipelines import OffPolicyPipeline as Pipeline
+from mava.utils.sebulba.rate_limiters import BlockingRatioLimiter, RateLimiter, SampleToInsertRatio
+from mava.utils.sebulba.utils import ParamsSource, RecordTimeTo
+from mava.wrappers.episode_metrics import get_final_step_metrics
+from mava.wrappers.gym import GymToJumanji
+
+
+def rollout(
+    key: chex.PRNGKey,
+    env: GymToJumanji,
+    config: DictConfig,
+    rollout_queue: Pipeline,
```

Comment:
In this case it's not a queue, because we're using a replay buffer. Can we rename this to pipeline